deep_quoridor/src/quoridor.py

```diff
@@ -0,0 +1,293 @@
+from dataclasses import dataclass
+from enum import IntEnum, unique
+
+import numpy as np
+
+type Position = np.ndarray[tuple[int, int], np.dtype[np.uint8]]  # (row, col)
+
+
+@unique
+class WallOrientation(IntEnum):
+    VERTICAL = 0
+    HORIZONTAL = 1
+
+
+class Action:
+    pass
+
+
+@dataclass
+class MoveAction(Action):
+    destination: Position
+
+
+@dataclass
+class WallAction(Action):
+    position: Position  # Start of wall
+    orientation: WallOrientation
+
+
+class Board:
+    """
+    Grid which has both walls and possible pawn positions.
+
+    Pawn positions are on even rows and columns, walls are on odd rows and columns. We use negative
+    values to represent free cells and walls, 0 for player 1, and 1 for player 2.
+    integers.
+    """
+
+    # Possible values for each cell in the grid.
+    OUT_OF_BOUNDS = -2  # Returned if a move is off of the board.
+    FREE = -1
+    # The player numbers start at 0 so that they can also be used as indices into the player positions array.
+    PLAYER1 = 0
+    PLAYER2 = 1
+    # Walls are stored as 10 + player number.We leave 2-9 free for implementing more than 2 players in the future.
+    PLAYER1_WALL = 10
+    PLAYER2_WALL = 11
+
+    def __init__(self, board_size: int, max_walls: int):
+        self.board_size = board_size
+        self.max_walls = max_walls
+
+        self._grid = np.full((board_size * 2 - 1, board_size * 2 - 1), Board.FREE, dtype=np.int8)
+        self._player_positions = np.array([(0, board_size // 2), (board_size - 1, board_size // 2)])
+        for player, position in enumerate(self._player_positions):
+            self._grid[*position * 2] = player
+        self._walls_remaining = [self.max_walls, self.max_walls]
+
+    def get_player_position(self, player: int) -> Position:
+        """
+        Get the position of the player's pawn.
+        """
+        try:
+            return self._player_positions[player]
+        except IndexError:
+            raise ValueError(f"Invalid player {player}")
+
+    def move_player(self, player: int, new_position: Position):
+        """
+        Move the player's pawn. Doesn't check if the move is valid according to game rules.
+        """
+        new_position = np.asarray(new_position, dtype=np.uint8)
+
+        try:
+            old_position = self._player_positions[player]
+        except IndexError:
+            raise ValueError(f"Invalid player {player}")
+
+        try:
+            self._grid[*old_position * 2] = Board.FREE
+            self._grid[*new_position * 2] = player
+            self._player_positions[player] = new_position
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+    def get_cell(self, position: Position) -> int:
+        """
+        Get the value of a "player cell" in the grid.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+        try:
+            return self._grid[*position * 2]
+        except IndexError:
```

Comment:
I prefer explicitly checking the bounds beforehand. Especially if we are going to return instead of reraise.